crashtracker/src/crash_handler.rs

```diff
@@ -66,89 +62,133 @@ fn make_receiver(
-    // Use the pipe to avoid secrets ending up on the commandline
-    writeln!(
-        receiver.stdin.as_ref().unwrap(),
-        "{}",
-        serde_json::to_string(&config)?
-    )?;
-    writeln!(
-        receiver.stdin.as_ref().unwrap(),
-        "{}",
-        serde_json::to_string(&metadata)?
-    )?;
+    update_metadata(metadata)?;
+    update_config(config)?;
+
     Ok(receiver)
 }
 
+/// Updates the crashtracker metadata for this process
+/// Metadata is stored in a global variable and sent to the crashtracking
+/// receiver when a crash occurs.
+///
+/// PRECONDITIONS:
+///     None
+/// SAFETY:
+///     Crash-tracking functions are not guaranteed to be reentrant.
+///     No other crash-handler functions should be called concurrently.
+/// ATOMICITY:
+///     This function uses a swap on an atomic pointer.
+pub fn update_metadata(metadata: CrashtrackerMetadata) -> anyhow::Result<()> {
+    let metadata_string = serde_json::to_string(&metadata)?;
+    let box_ptr = Box::into_raw(Box::new((metadata, metadata_string)));
+    let old = METADATA.swap(box_ptr, SeqCst);
+    if !old.is_null() {
+        // Safety: This can only come from a box above.
+        unsafe {
+            std::mem::drop(Box::from_raw(old));
+        }
+    }
+    Ok(())
+}
+
+/// Updates the crashtracker config for this process
+/// Config is stored in a global variable and sent to the crashtracking
+/// receiver when a crash occurs.
+///
+/// PRECONDITIONS:
+///     None
+/// SAFETY:
+///     Crash-tracking functions are not guaranteed to be reentrant.
+///     No other crash-handler functions should be called concurrently.
+/// ATOMICITY:
+///     This function uses a swap on an atomic pointer.
+pub fn update_config(config: CrashtrackerConfiguration) -> anyhow::Result<()> {
+    let config_string = serde_json::to_string(&config)?;
+    let box_ptr = Box::into_raw(Box::new((config, config_string)));
+    let old = CONFIG.swap(box_ptr, SeqCst);
+    if !old.is_null() {
+        // Safety: This can only come from a box above.
+        unsafe {
+            std::mem::drop(Box::from_raw(old));
+        }
+    }
+    Ok(())
+}
+
 pub fn setup_receiver(
-    config: &CrashtrackerConfiguration,
-    metadata: &CrashtrackerMetadata,
+    config: CrashtrackerConfiguration,
+    metadata: CrashtrackerMetadata,
 ) -> anyhow::Result<()> {
-    let new_receiver = make_receiver(config, metadata)?;
-    let old_receiver =
-        unsafe { std::mem::replace(&mut RECEIVER, GlobalVarState::Some(new_receiver)) };
+    let new_receiver = Box::into_raw(Box::new(make_receiver(config, metadata)?));
+    let old_receiver = RECEIVER.swap(new_receiver, SeqCst);
     anyhow::ensure!(
-        matches!(old_receiver, GlobalVarState::Unassigned),
-        "Error registering crash handler receiver: receiver already existed {old_receiver:?}"
+        old_receiver.is_null(),
+        "Error registering crash handler receiver: receiver already existed"
     );
```

Comment:
If the semantic here is to say an existing receiver already existing is an error, should we do a compare and swap on `RECEIVER` instead, expecting it to be `null`?

Right now it's kinda weird that we successfully set up the new receiver, and only then we return an error.

Reply:
I modified the semantics of registering a reciever